README.md

```diff
@@ -1,34 +1,332 @@
+
+    /// Free resource specified by given `resource` and `size`.
+    ///
+    /// # Arguments
+    ///
+    /// * `resource`: resource to be free.
+    /// * `size`: resource size of free request.
+    fn free(&mut self, resource: T, size: S);
+
+    /// Update resource, freeing the old one and allocating a new range.
+    ///
+    /// This is mostly used when guest allocated some resources or reprogrammed,
+    /// so the new allocated resources should be marked as used, and the preallocated
+    /// one should be freed (if has). 
+    ///
+    /// # Arguments
+    ///
+    /// * `old_resource`: old resource to be free or no such preallocated resources.
+    /// * `old_size`: old resource size of free request or no such preallocated resources.
+    /// * `resource`: resource to be free.
+    /// * `size`: resource size of free request.
+    fn update(&mut self, old_resource: Option<T>, old_size: Option<S>, new_resource: T, new_size: S);
+}
+```
+
+Different kinds of resources and its allocators should implement the three traits. Take
+address space resource for example.
+
+```rust
+impl Resource for GuestAddress {}
+
+// This should be implemented for u64.
+// Otherwise implementing for GuestUsize (aka. <GuestAddress as AddressValue>::V)
+// would impact other implementation such as instance id since compiler
+// doesn't know real type of V.
+impl ResourceSize for u64 {}
+
+pub struct AddressAllocator {
+    base: GuestAddress,
+    end: GuestAddress,
+    alignment: GuestUsize,
+    ranges: BTreeMap<GuestAddress, GuestUsize>,
+}
+
+impl ResourceAllocator<GuestAddress, GuestUsize> for AddressAllocator {
+    /// Allocates a range of addresses from the managed region.
+    fn allocate(
+        &mut self,
+        address: Option<GuestAddress>,
+        size: GuestUsize,
+    ) -> Result<GuestAddress> {...}
+
+    /// Free an already allocated address range.
+    fn free(&mut self, address: GuestAddress, size: GuestUsize) {...}
+
+    /// Update an address range with a new one.
+    fn update(&mut self, old_resource: Option<T>, old_size: Option<S>, new_resource: T, new_size: S) {...}
+```
+
+Another resource being used by VMM is unsigned integer resource,
+like IRQ numbers including legacy IRQ numbers and MSI/MSI-X vectors, device instance id.
+
+```rust
+impl Resource for u32 {}
+impl ResourceSize for u32 {}
+pub struct IdAllocator {
+    start: u32,
+    end: u32,
+    used: Vec<u32>,
+}
+impl ResourceAllocator<u32, u32> for IdAllocator {...}
+```
+
+### Design Note:
+
+VMM is responsible for system level resources allocation, and some principles and
+special cases should be taken into account.
+
+- Vmm probably need record which kinds of resources it has, like how many IRQ numbers, how
+many IO address ranges. 
+
+Let VMM design a `SystemAllocator` struct because only VMM knows exactly which resources it
+needs and whether it needs thread safe.
+
+Meanwhile, different resource instances are not suggested to being put together using like
```

Comment:
This is also one of the reasons why I think that we should not export a `ResourceAllocator` in the first place. We should make it hard to misuse the public interface of a crate.